tooling/debugger/src/context.rs

```diff
@@ -15,56 +16,138 @@ use nargo::errors::{ExecutionError, Location};
 use nargo::NargoError;
 use noirc_driver::DebugFile;
 
+use thiserror::Error;
+
 use std::collections::BTreeMap;
 use std::collections::{hash_set::Iter, HashSet};
 
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
+pub struct DebugLocation {
+    pub circuit_id: u32,
+    pub opcode_location: OpcodeLocation,
+}
+
+impl std::fmt::Display for DebugLocation {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let circuit_id = self.circuit_id;
+        match self.opcode_location {
+            OpcodeLocation::Acir(index) => write!(f, "{circuit_id}:{index}"),
+            OpcodeLocation::Brillig { acir_index, brillig_index } => {
+                write!(f, "{circuit_id}:{acir_index}.{brillig_index}")
+            }
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum DebugLocationFromStrError {
+    #[error("Invalid debug location string: {0}")]
+    InvalidDebugLocationString(String),
+}
+
+impl std::str::FromStr for DebugLocation {
+    type Err = DebugLocationFromStrError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts: Vec<_> = s.split(':').collect();
+
+        if parts.is_empty() || parts.len() > 2 {
```

Comment:
This feels like if could be cleaned up a little bit. We return an error here, panic inside of `parse_components`, and then do an `ok_or` that will never be hit because we only return `Some(..)` or panic.